### PR TITLE
[10.0][REF] refactor url generator

### PR DESCRIPTION
**Pull request by sebastienbeau**
- add generic method
- split the method to be able to customize easily the url key with the possibility to enjoy all the work that process record per lang and so enjoying the prefetching
- fix onchange issue by skipping compute when calling from an onchange

### CODE REVIEW

**Comment by sebastienbeau** [comment]
Hi @lmignon I would like to have your point of view on this PR.
Initialy my aim was not to have a generic implementation of the code for generation the url key, but only to be able to customize it easily (adding the brand in the url, adding the color, add some extra stuff...). Same for category, without re-writing the great logic of processing record per lang.
After spliting the method I was able to have the same generic method for product and categ so I move the logic in base_url.

File: base_url/models/abstract_url.py [hunk 1]
```diff
@@ -73,6 +74,34 @@ def _compute_is_urls_sync_required(self):
...
+    def _get_url_keyword(self):
+        self.ensure_one()
+        return [self.name]
+
+    def _post_process_url_key(self, key):
+        self.ensure_one()
+        return key
+
+    def _generic_compute_automatic_url_key(self):
+        if not self.ids:
```
Comment (lmignon) [inline base_url/models/abstract_url.py]: @sebastienbeau Your test is not right. self.ids could be a list with a mix of id and NewId. You must skip NewId when you loop on self

File: base_url/models/abstract_url.py [hunk 2]
```diff
@@ -73,6 +74,34 @@ def _compute_is_urls_sync_required(self):
         for record in self:
             record.is_urls_sync_required = True
 
+    def _get_url_keyword(self):
```
Comment (lmignon) [inline base_url/models/abstract_url.py]: @sebastienbeau `s/get_url_keyword/get_url_keywrods`

Comment (lmignon) [reply base_url/models/abstract_url.py]: @sebastienbeau Some docstring explaining the purpose of each methods into abstract_url would be nice.

File: base_url/tests/test_abstract_url.py [hunk 3]
```diff
@@ -118,3 +118,10 @@ def test_write_inactive(self):
         ) as mocked_redirect:
             my_partner.active = False
             mocked_redirect.assert_called_once()
+
+    def test_onchange_do_nothing(self):
+        # calling onchange should not fail and should not return an url
```
Comment (lmignon) [inline base_url/tests/test_abstract_url.py:123]: @sebastienbeau You should add a tests for the oposite to ensure that an onchange on an existing record update the url

**Review by lmignon — CHANGES_REQUESTED** [review]
Some comments but it's indeed more generic

File: base_url/tests/test_abstract_url.py (hunk 3 above)
Comment (sebastienbeau) [reply base_url/tests/test_abstract_url.py:123]: In fact onchange are always played with NewID (even if the record exist). Regarding the case of testing the _generic_compute_automatic_url_key it's already done with the other test (that check the url generated automatically)

**Review by lmignon — APPROVED** [review]

**Pull request merged**